File: java/org/servalproject/servaldna/AbstractId.java
/**
 * Copyright (C) 2011 The Serval Project
 *
 * This file is part of Serval Software (http://www.servalproject.org)
 *
 * Serval Software is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This source code is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this source code; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

package org.servalproject.servaldna;

import java.nio.BufferUnderflowException;
import java.nio.ByteBuffer;

public abstract class AbstractId {

	abstract int getBinarySize();

	public static class InvalidHexException extends Exception {
		private static final long serialVersionUID = 1L;

		private InvalidHexException(AbstractId id, String message) {
			super(id.getClass().getName() + ": " + message);
		}
	}

	public static class InvalidBinaryException extends Exception {
		private static final long serialVersionUID = 1L;

		private InvalidBinaryException(AbstractId id, String message) {
			super(id.getClass().getName() + ": " + message);
		}
	}

	private final byte[] binary;

	public AbstractId(String hex) throws InvalidHexException {
		if (hex==null)
			throw new InvalidHexException(this, "null is not a invalid hex value");
		if (hex.length() != getBinarySize()*2)
			throw new InvalidHexException(this, "invalid length " + hex.length() + " (should be " + (getBinarySize() * 2) + ") of '" + hex + "'");
		binary = new byte[getBinarySize()];
		int j = 0;
		for (int i = 0; i != binary.length; i++) {
			int d1 = Character.digit(hex.charAt(j++), 16);
			int d2 = Character.digit(hex.charAt(j++), 16);
			if (d1 == -1 || d2 == -1)
				throw new InvalidHexException(this, "non-hex digit in '" + hex + "'");
			binary[i] = (byte) ((d1 << 4) | d2);
		}
	}

	public AbstractId(ByteBuffer b) throws InvalidBinaryException {
		this.binary = new byte[getBinarySize()];
		try {
			b.get(this.binary);
		}
		catch (BufferUnderflowException e) {
			throw new InvalidBinaryException(this, "not enough bytes (expecting " + getBinarySize() + ")");
		}
	}

	public AbstractId(byte[] binary) throws InvalidBinaryException {
		if (binary.length != getBinarySize())
			throw new InvalidBinaryException(this, "invalid number of bytes (" + binary.length + "), should be " + getBinarySize());
		this.binary = binary;
	}

	@Override
	public boolean equals(Object other) {
		// must be the exact same type with the same binary contents to be considered equal
		if (other==null)
			return false;
		if (other==this)
			return true;
		if (other.getClass() == this.getClass()) {
			AbstractId oBinary = (AbstractId) other;
			for (int i = 0; i < this.binary.length; i++)
				if (this.binary[i] != oBinary.binary[i])
					return false;
			return true;
		}
		return false;
	}

	@Override
	public int hashCode() {
		int hashCode = 0;
		for (int i = 0; i < this.binary.length; i++)
			hashCode = (hashCode << 8 | hashCode >>> 24) ^ this.binary[i];
		return hashCode;
	}

	public void toByteBuffer(ByteBuffer buff){
		buff.put(this.binary);
	}

	public String toHex(int offset, int len) {
		StringBuilder sb = new StringBuilder();
		for (int i = offset; i < offset + len && i < binary.length; i++) {
			sb.append(Character.forDigit(((binary[i]) & 0xf0) >> 4, 16));
			sb.append(Character.forDigit((binary[i]) & 0x0f, 16));
		}
		return sb.toString().toUpperCase();
	}

	public String toHex(int len) {
		return toHex(0, len);
	}

	public String toHex() {
		return toHex(0, binary.length);
	}

	public String abbreviation() {
		return toHex(0, 4);
	}


	@Override
	public String toString() {
		return toHex();
	}

}

File: swift-client-api/package/Sources/abstract_id.swift
/*
Serval DNA Swift API
Copyright (C) 2016 Flinders University

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
*/

import Foundation

/* All the Serval Id types (SID, BundleId, etc.) are binary blobs that are
* typically represented in hexadecimal text format.  Some also have an
* "abbreviated" format.
*/

public protocol AbstractId : Hashable, CustomStringConvertible {
    init(fromBinary: [UInt8])
    init?(fromHex: String)
    var binary : [UInt8] { get }
    var hexUpper : String { get }
}

public protocol AbbreviatedId : AbstractId {
    var abbreviation : String { get }
}

/* Implementations of AbstractId that specialise GenericIdImplementation must
* also implement the ConcreteId protocol, which allows users to discover
* properties of the specialised type without instantiating it.
*/

public protocol ConcreteId {
    static var byteCount : Int { get }
    static var mimeType : String { get }
}

/* All Serval Id types are specialisations of this generic implementation,
* which stores the Id's value as an array of bytes, and derives all other
* properties from that representation.
*/

public class GenericIdImplementation : AbstractId {
    public let binary : [UInt8]

    public static func == (lhs: GenericIdImplementation, rhs: GenericIdImplementation) -> Bool {
        return lhs.binary == rhs.binary
    }

    public var hashValue : Int {
        return self.binary.map { $0.hashValue }.reduce(0) {
            (($0 << 8) | Int(UInt($0) >> UInt((MemoryLayout<Int>.size - 1) * 8))) ^ $1
        }
    }

    internal init(fill: UInt8) {
        self.binary = Array(repeating: fill, count: (type(of:self) as! ConcreteId.Type).byteCount)
    }

    public required init(fromBinary binary: [UInt8]) {
        self.binary = binary
        assert(self.binary.count == (type(of:self) as! ConcreteId.Type).byteCount)
    }

    public required init?(fromHex hex: String) {
        var binary = Array(repeating: UInt8(0), count: (type(of:self) as! ConcreteId.Type).byteCount)
        var index = hex.startIndex
        for i in 0 ..< binary.count {
            guard let next = hex.index(index, offsetBy: 2, limitedBy: hex.endIndex) else {
                return nil
            }
            let digits = hex.substring(with: index ..< next)
            guard let byte = UInt8(digits, radix: 16) else {
                return nil
            }
            binary[i] = byte
            index = next
        }
        guard index == hex.endIndex else {
            return nil
        }
        self.binary = binary
    }

    public var hexUpper : String {
        return self.binary.map { String(format: "%02hhX", $0) }.joined()
    }

    internal func hexUpper(digitCount: Int) -> String {
        let byteCount = (digitCount + 1) / 2
        assert(byteCount <= self.binary.count)
        let hex = self.binary[0 ..< byteCount].map { String(format: "%02hhX", $0) }.joined()
        return digitCount == byteCount * 2 ? hex : hex.substring(with: hex.startIndex ..< hex.index(hex.endIndex, offsetBy: -1))
    }

    public var description : String {
        return "\(type(of:self))(fromHex: \"\(self.hexUpper)\")"
    }
}
